klippy/mcu.py

```diff
@@ -246,32 +256,47 @@ def _build_config(self):
+        self._set_cmd = self._mcu.lookup_command_tag(
+            "queue_digital_out oid=%c clock=%u on_ticks=%u")
         self._set_cycle_ticks = self._mcu.lookup_command(
             "set_digital_out_pwm_cycle oid=%c cycle_ticks=%u", cq=cmd_queue)
     def set_pwm(self, print_time, value, cycle_time=None):
-        clock = self._mcu.print_time_to_clock(print_time)
+        req_clock = self._mcu.print_time_to_clock(print_time)
+        # FIXME: let sync_channel replace uncommitted values
+        #        so that this is not necessary
         minclock = self._last_clock
-        self._last_clock = clock
+        clock = max(req_clock, self._last_clock + self._min_clock_diff)
         if self._invert:
             value = 1. - value
         if self._hardware_pwm:
             v = int(max(0., min(1., value)) * self._pwm_max + 0.5)
-            self._set_cmd.send([self._oid, clock, v],
-                               minclock=minclock, reqclock=clock)
-            return
-        # Soft pwm update
-        if cycle_time is None:
-            cycle_time = self._cycle_time
-        cycle_ticks = self._mcu.seconds_to_clock(cycle_time)
-        if cycle_ticks != self._last_cycle_ticks:
-            self._set_cycle_ticks.send([self._oid, cycle_ticks],
-                                       minclock=minclock, reqclock=clock)
-            self._last_cycle_ticks = cycle_ticks
-        on_ticks = int(max(0., min(1., value)) * float(cycle_ticks) + 0.5)
-        self._set_cmd.send([self._oid, clock, on_ticks],
-                           minclock=minclock, reqclock=clock)
+            # queue_pwm_out oid=%c clock=%u value=%hu
+            data = (self._set_cmd, self._oid, clock & 0xFFFFFFFF, v)
+        else:
+            # Soft pwm update
+            if cycle_time is None:
+                cycle_time = self._cycle_time
+            cycle_ticks = self._mcu.seconds_to_clock(cycle_time)
+            if cycle_ticks != self._last_cycle_ticks:
+                self._set_cycle_ticks.send([self._oid, cycle_ticks],
+                                           minclock=minclock, reqclock=clock)
+                self._last_cycle_ticks = cycle_ticks
+                self._min_clock_diff = cycle_ticks
+                # FIXME: make sure new cycle_ticks is applied
+                # _before_ new on_ticks arrive
+                clock = self._mcu.print_time_to_clock(print_time + 0.100)
+            on_ticks = int(max(0., min(1., value)) * float(cycle_ticks) + 0.5)
+            # queue_digital_out oid=%c clock=%u on_ticks=%u
+            data = (self._set_cmd, self._oid, clock & 0xFFFFFFFF, on_ticks)
+
+        self._ffi_lib.sync_channel_queue_msg(self._sync_channel,
+                                             data, len(data), clock)
+
+        # TODO: Is here actually `register_async_callback` needed?
+        self.reactor.register_callback(
+            lambda ev: self._th.note_synchronous_command(print_time))
```

Comment:
@KevinOConnor If heater-code (background task) calls this, is actually `register_async_callback` needed?
It worked both ways for me.